src/Protocol/ForgeHandshake.cpp

```diff
@@ -0,0 +1,367 @@
+
+	Json::Value Modinfo;
+	Modinfo["type"] = "FML";
+
+	Json::Value ModList(Json::arrayValue);
+	for (auto & item: Mods)
+	{
+		Json::Value Mod;
+		Mod["modid"] = item.first;
+		Mod["version"] = item.second;
+		ModList.append(Mod);
+	}
+	Modinfo["modList"] = ModList;
+
+	a_ResponseValue["modinfo"] = Modinfo;
+}
+
+
+
+
+
+void cForgeHandshake::BeginForgeHandshake(const AString & a_Name, const AString & a_UUID, const Json::Value & a_Properties)
+{
+	ASSERT(m_IsForgeClient);
+
+	m_Name = a_Name;
+	m_UUID = a_UUID;
+	m_Properties = a_Properties;
+
+	static const std::array<AString, 5> Channels{{ "FML|HS", "FML", "FML|MP", "FML", "FORGE" }};
+	AString ChannelsString;
+	for (auto & Channel: Channels)
+	{
+		ChannelsString.append(Channel);
+		ChannelsString.push_back('\0');
+	}
+
+	m_Client->SendPluginMessage("REGISTER", ChannelsString);
+	SendServerHello();
+}
+
+
+
+
+
+void cForgeHandshake::SendServerHello()
+{
+	AString Message;
+	// Discriminator | Byte | Always 0 for ServerHello
+	Message.push_back(Discriminator::ServerHello);
+	// FML protocol Version | Byte | Determined from NetworkRegistery. Currently 2.
+	Message.push_back('\2');
+	// Dimension TODO
+	Message.push_back('\0');
+	Message.push_back('\0');
+	Message.push_back('\0');
+	Message.push_back('\0');
+
+	m_Client->SendPluginMessage("FML|HS", Message);
+}
+
+
+
+
+
+AStringMap cForgeHandshake::ParseModList(const char * a_Data, size_t a_Size)
+{
+	AStringMap Mods;
+
+	if (a_Size < 4)
+	{
+		SetError(Printf("ParseModList invalid packet, missing length (size = " SIZE_T_FMT ")", a_Size));
+		return Mods;
+	}
+
+	cByteBuffer Buf(a_Size);
+	Buf.Write(a_Data, a_Size);
+	UInt32 NumMods;
+	if (!Buf.ReadVarInt32(NumMods))
+	{
+		SetError("ParseModList failed to read mod count");
+		return Mods;
+	}
+
+	for (UInt32 i = 0; i < NumMods; ++i)
+	{
+		AString Name, Version;
+		if (!Buf.ReadVarUTF8String(Name))
+		{
+			SetError(Printf("ParseModList failed to read mod name at i = %d", i));
+			break;
+		}
+		if (!Buf.ReadVarUTF8String(Version))
+		{
+			SetError(Printf("ParseModList failed to read mod version at i = %d", i));
+			break;
+		}
+		Mods.insert({Name, Version});
+	}
+
+	return Mods;
+}
+
+
+
+
+void cForgeHandshake::HandleClientHello(cClientHandle * a_Client, const char * a_Data, size_t a_Size)
+{
+	if (a_Size == 2)
+	{
+		int FmlProtocolVersion = a_Data[1];
+		LOGD("Received ClientHello with FML protocol version %d", FmlProtocolVersion);
+		if (FmlProtocolVersion != 2)
+		{
+			SetError(Printf("Unsupported FML client protocol version received in ClientHello: %d", FmlProtocolVersion));
+		}
+	}
+	else
+	{
+		SetError(Printf("Received unexpected length of ClientHello: " SIZE_T_FMT, a_Size));
+	}
+}
+
+
+
+
+void cForgeHandshake::HandleModList(cClientHandle * a_Client, const char * a_Data, size_t a_Size)
+{
+	LOGD("Received ModList");
+
+	auto ClientMods = ParseModList(a_Data + 1, a_Size - 1);
+	AString ClientModsString;
+	for (auto & item: ClientMods)
+	{
+		ClientModsString.append(item.first);
```

Comment:
Would
```
ClientModsString.append(Printf("%s@%s, ", item.first, item.second))
```
work? It might be slightly slower, but it'd be one line instead of four.

Reply:
Slight change to pass C strings, cleaner and works well:

```c++
AppendPrintf(ClientModsString, "%s@%s, ", item.first.c_str(), item.second.c_str());
```